daemon/daemon_main.go

```diff
@@ -823,16 +833,61 @@ func initEnv(cmd *cobra.Command) {
 		log.WithError(err).Fatal("Invalid sidecar-istio-proxy-image regular expression")
 		return
 	}
+
+	if option.Config.PolicyEnforcementInterface == "" {
+		option.Config.PolicyEnforcementInterface = viper.GetString("policy-enforcement-interface")
+	}
+
+	if !option.Config.PolicyEnforcementCleanUp {
+		option.Config.PolicyEnforcementCleanUp = viper.GetBool("policy-enforcement-clean-up")
+	}
+}
+
+// setHostDeviceWhenReady sets the given ifaceName as the HostDevice. If
+// ifaceName is not found, then it will wait forever until the device is
+// created before it assign it to the hostDevice.
+func setHostDeviceWhenReady(ifaceName string) {
+	for i := 0; ; i++ {
+		if i%10 == 0 {
+			log.WithField("interface", ifaceName).
+				Info("Waiting for the underlying interface to be initialized with containers")
+		}
+		_, err := netlink.LinkByName(ifaceName)
+		if err == nil {
+			log.WithField("interface", ifaceName).
+				Info("Underlying interface initialized with containers!")
+			break
+		}
+		time.Sleep(time.Second)
+	}
+	defaults.HostDevice = ifaceName
 }
 
 func runDaemon() {
 	log.Info("Initializing daemon")
+
+	// Since flannel doesn't create the cni0 interface until the first container
+	// is initialized we need to wait until it is initialized so we can attach
+	// the BPF program to it. If Cilium is running as a Kubernetes DaemonSet,
+	// there is also a script waiting for the interface to be created.
+	if option.Config.IsPolicyEnforcementInterfaceSet() {
+		setHostDeviceWhenReady(option.Config.PolicyEnforcementInterface)
+	}
+
 	d, restoredEndpoints, err := NewDaemon()
 	if err != nil {
 		log.WithError(err).Fatal("Error while creating daemon")
 		return
 	}
 
+	if option.Config.IsPolicyEnforcementInterfaceSet() && option.Config.PolicyEnforcementCleanUp {
+		pidfile.HandleCleanup(cleanUPWg, cleanUPSig, func() {
+			d.compilationMutex.Lock()
+			loader.DeleteDatapath(context.Background(), defaults.HostDevice, "egress")
```

Comment:
Does this mean that we remove Cilium policy enforcement during Cilium restart / upgrade?

Reply:
Yes, check https://github.com/cilium/cilium/pull/6267#discussion_r238885829